{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b34e2840-69b8-45af-8ccd-7d2fc8e0d6c9",
   "metadata": {},
   "source": [
    "# ChromMovie tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8827408-fa92-4668-b10f-35a4a0d9dfa4",
   "metadata": {},
   "source": [
    "Author: Krzysztof H. Banecki, 2025"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1478af64-3d2f-43be-b86e-a9b6cbae2d9a",
   "metadata": {},
   "source": [
    "## Cloning the repo and installing dependencies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f881698-8789-48c4-a0c2-874875fb5654",
   "metadata": {},
   "source": [
    "If running for the first time, please close ChromMovie repository:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a558948b-55e5-48aa-abd0-8788ebc433af",
   "metadata": {},
   "outputs": [],
   "source": [
    "!git clone https://github.com/SFGLab/ChromMovie"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64e87f2d-e339-43bc-9150-6a0516fef00b",
   "metadata": {},
   "source": [
    "And install the requirements (in case of the problems with numpy, please just restart the session and rerun the cell below):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b859dc6-7ac5-41fd-acc5-cd9a2a595b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ChromMovie\n",
    "!pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe8ed689-34d8-41ab-af37-7c528c65adf5",
   "metadata": {},
   "source": [
    "## Setting up simulation parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06188af2-ae3a-487c-95ed-794c9ed55685",
   "metadata": {},
   "source": [
    "All of the simulation parameters are stored in `YAML` files (e.g. `config.yaml`).\n",
    "\n",
    "When specifying the simulation parameters in your config file, mind the following parameters:\n",
    "\n",
    "* Input and output directories:\n",
    "```\n",
    "  input: examples/example1_cell_cycle\n",
    "  output: results\n",
    "```\n",
    "If output directory does not exist, it is going to be created. Here we will use the default example dataset from [1].\n",
    "\n",
    "* Reference genome:\n",
    "```\n",
    "    genome: hg38\n",
    "```\n",
    "Specify the reference genome of your input data. Available reference genomes include: `hg19`, `hg38`, `mm10`, `GRCm39` and `GRCg6a`, although one can add another reference just by adding appropriate chromosome sizes file in `chrom_sizes` folder.\n",
    "\n",
    "* Simulation platform:\n",
    "```\n",
    "    platform: CPU\n",
    "```\n",
    "Available platoforms: `CPU`, `CUDA` and `OpenCL`. Here, we'll use standard `CPU` platform.\n",
    "\n",
    "* Modeling resolutions:\n",
    "```\n",
    "    resolutions: 5,2\n",
    "```\n",
    "Resolutions to be used for hierarchical modeling. Expected to be in the form of comma separated integer of float numbers in the units of Mb. In the above example simulation will employ 5Mb followed by 2Mb resolution simulation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb4e753f-b822-4228-8e4e-845ca84ea525",
   "metadata": {},
   "source": [
    "## Running ChromMovie simulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a7d268-e254-41ac-a82d-26b91481c044",
   "metadata": {},
   "source": [
    "After specifying all the relevant simulation parameters in the `config.yaml` file, the simulation is ready to deploy:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93540071-1e92-441a-a662-1cc37024cc9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!python3 -m ChromMovie -i config.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af22b64-66f1-4656-9b11-0950ed105c67",
   "metadata": {},
   "source": [
    "## Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b7b44ae-8364-4ba6-8c42-e98e71dd4d4e",
   "metadata": {},
   "source": [
    "The simulation results should appear in the specified `output` folder. The subfolder `frames_cif` contains all of the modeled structures in `.cif` format. The structures can be viewed directly for example using the software UCSF Chimera [2].\n",
    "\n",
    "Detailed information about other output files is contained in the `README.md` file in the Github repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bbd180a-1e7f-47e0-b8a0-37558381f19d",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95cf0b44-7cbb-44f5-bbef-8649df668ac0",
   "metadata": {},
   "source": [
    "[1] Chai, H., Huang, X., Xiong, G. et al. Tri-omic single-cell mapping of the 3D epigenome and transcriptome in whole mouse brains throughout the lifespan. Nat Methods 22, 994â€“1007 (2025). https://doi.org/10.1038/s41592-025-02658-7\n",
    "\n",
    "[2] UCSF Chimera--a visualization system for exploratory research and analysis. Pettersen EF, Goddard TD, Huang CC, Couch GS, Greenblatt DM, Meng EC, Ferrin TE. J Comput Chem. 2004 Oct;25(13):1605-12. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc37843-56d9-424b-ac69-40c2441dddc5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
